src/cockpit/channels/filesystem.py

```diff
@@ -77,39 +79,38 @@ def do_open(self, options):
             self.close()
 
 
-class FsReadChannel(Channel):
+class FsReadChannel(GeneratorChannel):
     payload = 'fsread1'
 
-    def do_open(self, options):
-        self.ready()
-        try:
-            logger.debug('Opening file "%s" for reading', options['path'])
-            try:
-                with open(options['path'], 'rb') as filep:
-                    buf = os.stat(filep.fileno())
-                    tag = tag_from_stat(buf)
-                    if max_read_size := options.get('max_read_size'):
-                        if buf.st_size > max_read_size:
-                            raise ChannelError('too-large')
-
-                    data = filep.read()
-            except FileNotFoundError:
-                self.close(tag='-')
-                return
-            except PermissionError:
-                raise ChannelError('access-denied')
-            except OSError:
-                raise ChannelError('internal-error')
+    def do_yield_data(self, size_hint: int, options: Dict[str, object]) -> GeneratorChannel.DataGenerator:
+        binary = options.get('binary', False)
+        max_read_size = options.get('max_read_size')
+
+        logger.debug('Opening file "%s" for reading', options['path'])
 
-            if 'binary' not in options:
-                data = data.replace(b'\0', b'').decode('utf-8', errors='ignore').encode('utf-8')
+        try:
+            with open(options['path'], 'rb') as filep:
+                buf = os.stat(filep.fileno())
+                if max_read_size is not None and buf.st_size > max_read_size:
+                    raise ChannelError('too-large')
+
+                while True:
+                    data = filep.read1(size_hint)
+                    if data == b'':
+                        break
+                    logger.debug('  ...sending %d bytes', len(data))
+                    if not binary:
+                        data = data.replace(b'\0', b'').decode('utf-8', errors='ignore').encode('utf-8')
+                    yield data
+
+            return {'tag': tag_from_stat(buf)}
 
-            logger.debug('  ...sending %d bytes', len(data))
-            self.send_data(data)
         except FileNotFoundError:
-            logger.debug('  ...file not found!')
-        self.done()
-        self.close(tag=tag)
+            return {'tag': '-'}
+        except PermissionError:
+            raise ChannelError('access-denied')
+        except OSError:
+            raise ChannelError('internal-error')
```

Comment:
While you are at it: Can we send the contents of the OSError as message in the channel error? or at least `logger` it? Ignoring it entirely makes things hard to debug. This is not a regression, thus can also be a follow-up.

Reply:
Sent to PR #18285